File: cs/auth/2.private/hyper_id_sdk_impl.cs
using HyperId.SDK;
using HyperId.SDK.Authorization;
using HyperId.SDK.KYC;
using HyperId.SDK.MFA;
using HyperId.SDK.Storage;
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using System.Threading;

namespace HyperId.Private
{
    internal class HyperIDSDKImpl : IHyperIDSDK
    {
        private AuthSDKImpl         auth        = new AuthSDKImpl();
        private IHyperIDSDKMFA?     mfa;
        private IHyperIDSDKKyc?     kyc;
        private IHyperIDSDKStorage? storage;

        async Task IHyperIDSDK.InitAsync([NotNull] ProviderInfo providerInfo,
            [NotNull] ClientInfo clientInfo,
            [AllowNull] string? authRestoreInfo,
            CancellationToken cancellationToken)
        {
            await auth.InitAsync(providerInfo,
                clientInfo,
                authRestoreInfo,
                cancellationToken);
        }

        void IHyperIDSDK.Done()
        {
            auth.Done();
        }

        IHyperIDSDKAuth IHyperIDSDK.GetAuth()
        {
            return auth;
        }

        IHyperIDSDKMFA IHyperIDSDK.GetMFA()
        {
            if (mfa == null)
            {
                mfa = new MFASDKImpl((IHyperIDSDKAuthRestApi)auth);
            }
            return mfa;
        }

        IHyperIDSDKKyc IHyperIDSDK.GetKYC()
        {
            if(kyc == null)
            {
                kyc = new KycSDKImpl((IHyperIDSDKAuthRestApi)auth);
            }
            return kyc;
        }

        IHyperIDSDKStorage IHyperIDSDK.GetStorage()
        {
            if(storage == null)
            {
                storage = new StorageSDKImpl((IHyperIDSDKAuthRestApi)auth);
            }
            return storage;
        }

        public string? GetAuthRestoreInfo()
        {
            return auth.RefreshToken();
        }
    }
}//namespace HyperId.Private

File: android/hyperid_sdk/src/main/java/ai/hyper_id/sdk/internal/hyper_id_sdk_impl.kt
package ai.hyper_id.sdk.internal

import ai.hyper_id.sdk.IHyperIdSDK
import ai.hyper_id.sdk.api.auth.IHyperIDSDKAuth
import ai.hyper_id.sdk.api.auth.model.ClientInfo
import ai.hyper_id.sdk.api.auth.model.HyperIdProvider
import ai.hyper_id.sdk.api.kyc.IHyperIDSDKKYC
import ai.hyper_id.sdk.api.mfa.IHyperIdSDKMFA
import ai.hyper_id.sdk.api.storage.IHyperIDSDKStorage
import ai.hyper_id.sdk.internal.auth.HyperIDSDKAuthImpl
import ai.hyper_id.sdk.internal.kyc.HyperIDSDKKYCImpl
import ai.hyper_id.sdk.internal.mfa.HyperIdMFAImpl
import ai.hyper_id.sdk.internal.storage.HyperIDSDKStorageImpl

//**************************************************************************************************
//	HyperIdSDKImpl
//--------------------------------------------------------------------------------------------------
class HyperIdSDKImpl : IHyperIdSDK
{
	private val auth									= HyperIDSDKAuthImpl()
	private var kyc 		: IHyperIDSDKKYC?			= null
	private var mfa 		: IHyperIdSDKMFA?			= null
	private var storage 	: IHyperIDSDKStorage?		= null

	//==================================================================================================
	//	init
	//--------------------------------------------------------------------------------------------------
	override fun init(provider			: HyperIdProvider,
					  clientInfo		: ClientInfo,
					  authRestoreInfo	: String?,
					  completeListener	: IHyperIdSDK.IRequestResultListener)
	{
		auth.init(provider,
				  clientInfo,
				  authRestoreInfo,
				  completeListener)
	}
	//==================================================================================================
	//	done
	//--------------------------------------------------------------------------------------------------
	override fun done()
	{
		kyc			= null
		mfa			= null
		storage		= null

		auth.done()
	}

	override fun getAuth()			: IHyperIDSDKAuth			= auth
	override fun getMFA()			: IHyperIdSDKMFA			= mfa			?: HyperIdMFAImpl(auth)
	override fun getKYC()			: IHyperIDSDKKYC			= kyc			?: HyperIDSDKKYCImpl(auth)
	override fun getStorage()		: IHyperIDSDKStorage		= storage		?: HyperIDSDKStorageImpl(auth)
}
